### PR TITLE
Filter and calculate transforms as an array. Part of #2005

### PR DESCRIPTION
**Pull request by domoritz**

### CODE REVIEW

File: src/compile/data/nullfilter.ts [hunk 1]
```diff
@@ -60,7 +57,7 @@ export const nullFilter: DataComponentCompiler<Dict<FieldDef>> = {
       const childDataComponent = child.component.data;
       if (model.compatibleSource(child) && !differ<FieldDef>(childDataComponent.nullFilter, nullFilterComponent)) {
         extend(nullFilterComponent, childDataComponent.nullFilter);
-        delete childDataComponent.nullFilter;
+        childDataComponent.nullFilter = {};
```
Comment (kanitw) [inline src/compile/data/nullfilter.ts]: Can you explain why you need to change this line?

File: src/compile/data/nullfilter.ts [hunk 2]
```diff
@@ -17,12 +17,9 @@ const DEFAULT_NULL_FILTERS = {
...
 /** Return Hashset of fields for null filtering (key=field, value = true). */
 function parse(model: Model): Dict<FieldDef> {
-  const filterInvalid = model.filterInvalid();
-
   return model.reduceFieldDef(function(aggregator: Dict<FieldDef>, fieldDef: FieldDef) {
-    if (fieldDef.field !== '*') { // Ignore * for count(*) fields.
-      if (filterInvalid ||
-        (filterInvalid === undefined && fieldDef.field && DEFAULT_NULL_FILTERS[fieldDef.type])) {
+    if (fieldDef.aggregate !== 'count') { // Ignore * for count(*) fields.
+      if (fieldDef.field && DEFAULT_NULL_FILTERS[fieldDef.type]) {
```
Comment (kanitw) [inline src/compile/data/nullfilter.ts]: It seems like we always filterInvalid for now.  If you wanna keep the functionality, at least add it to config to allow disabling?

File: src/compile/data/source.ts [hunk 3]
```diff
@@ -38,16 +37,16 @@ export namespace source {
...
-            {type: formatType ? formatType : defaultExtension},
-            dataFormat.property ? {property: dataFormat.property} : {},
-            // Feature and mesh are two mutually exclusive properties
-            dataFormat.feature ?
-              {feature : dataFormat.feature} :
-            dataFormat.mesh ?
-              {mesh : dataFormat.mesh} :
-              {}
-          );
+        extend(
```
Comment (kanitw) [inline src/compile/data/source.ts]: the new tabbing is weird. 

How about just using the ... syntax.

const {property, feature, mesh} = dataFormat;

sourceData.format = {
  type: formatType ? formatType : defaultExtension,
  ...(property ? {property} : {}),
  ...(feature ? {feature} : {}),
  ...(mesh ? {mesh} : {}),
}

File: src/transform.ts [hunk 4]
```diff
@@ -1,33 +1,29 @@
...
-  filter?: string | Filter | (string | Filter)[];
-  /**
-   * Whether to filter invalid values (`null` and `NaN`) from the data. By default (`undefined`), only quantitative and temporal fields are filtered. If set to `true`, all data items with null values are filtered. If `false`, all data items are included.
-   */
-  filterInvalid?: boolean;
-  /**
-   * Calculate new field(s) using the provided expresssion(s). Calculation are applied before filter.
-   */
-  calculate?: Formula[];
+  filter: Filter | Filter[];
```
Comment (kanitw) [inline src/transform.ts]: I thought you wanna get rid of `Filter[]`.  (This could become a separate PR though.)

File: src/transform.ts [hunk 5]
```diff
@@ -1,33 +1,29 @@
...
+export interface CalculateTransform {
   /**
-   * The field in which to store the computed formula value.
+   * A string containing an expression for the formula. Use the variable `datum` to to refer to the current data object.
    */
-  as: string;
+  calculate: string;
   /**
-   * A string containing an expression for the formula. Use the variable `datum` to to refer to the current data object.
+   * The field in which to store the computed formula value.
```
Comment (kanitw) [inline src/transform.ts]: The field for storing the computed formula value.

File: src/transform.ts [hunk 6]
```diff
@@ -1,33 +1,29 @@
...
- */
-export interface Formula {
+export function isFilter(t: Transform): t is FilterTransform {
+  return t['filter'] !== undefined;
+}
+
+export interface CalculateTransform {
   /**
-   * The field in which to store the computed formula value.
+   * A string containing an expression for the formula. Use the variable `datum` to to refer to the current data object.
```
Comment (kanitw) [inline src/transform.ts]: containing a Vega Expression formula

File: src/compile/data/nullfilter.ts (hunk 2 above)
Comment (kanitw) [inline src/compile/data/nullfilter.ts]: Another possibly idea is that we leave them in the array.  
If they use it twice, then they are just stupid.

File: src/transform.ts (hunk 4 above)
Comment (kanitw) [inline src/transform.ts]: I think for Voyager, the `Filter[]` is a little handy though.

File: src/compile/data/nullfilter.ts (hunk 2 above)
Comment (domoritz) [inline src/compile/data/nullfilter.ts]: We only filter Q and T, which is a sensible default.

File: src/compile/data/nullfilter.ts (hunk 1 above)
Comment (domoritz) [inline src/compile/data/nullfilter.ts]: I was expecting nullFilter to be always defined. If I don't do this, `Object.keys` breaks.

File: src/transform.ts (hunk 4 above)
Comment (domoritz) [inline src/transform.ts]: It is simple so I left it.

File: src/compile/data/nullfilter.ts (hunk 2 above)
Comment (kanitw) [inline src/compile/data/nullfilter.ts]: "sensible default" means users should be able to override?

Comment (domoritz) [inline src/compile/data/nullfilter.ts]: Will fix it.

**Comment by domoritz** [comment]
Done.

File: src/compile/data/nullfilter.ts (hunk 2 above)
Comment (domoritz) [inline src/compile/data/nullfilter.ts]: > "sensible default" means users should be able to override?

Actually, no. You cannot disable the filtering. We always filter nulls when we aggregate.

Comment (kanitw) [inline src/compile/data/nullfilter.ts]: But it's terrible if there is no way to show null values even for raw plots.

Comment (domoritz) [inline src/compile/data/nullfilter.ts]: I see. I thought we didn't support it in the past but we did. Fixed it.

**Pull request merged**